Sources/FoundationEssentials/Data/Data+Writing.swift

```diff
@@ -387,44 +389,112 @@ private func writeToFileAux(path inPath: PathOrURL, buffer: UnsafeRawBufferPoint
 
-        // Cleanup temporary directory
-        defer { cleanupTemporaryDirectory(at: temporaryDirectoryPath) }
+    // Cleanup temporary directory
+    defer { cleanupTemporaryDirectory(at: temporaryDirectoryPath) }
 
-        guard fd >= 0 else {
-            throw CocoaError.errorWithFilePath(inPath, errno: errno, reading: false)
+    guard fd >= 0 else {
+        throw CocoaError.errorWithFilePath(inPath, errno: errno, reading: false)
+    }
+
+    defer { if fd >= 0 { _close(fd) } }
+
+    let callback = (reportProgress && Progress.current() != nil) ? Progress(totalUnitCount: Int64(buffer.count)) : nil
+
+    do {
+        try write(buffer: buffer, toFileDescriptor: fd, path: inPath, parentProgress: callback)
+    } catch {
+        try auxPath.withNTPathRepresentation { pwszAuxPath in
+            _ = DeleteFileW(pwszAuxPath)
         }
 
-        defer { if fd >= 0 { _close(fd) } }
+        if callback?.isCancelled ?? false {
+            throw CocoaError(.userCancelled)
+        } else {
+            throw CocoaError.errorWithFilePath(inPath, errno: errno, reading: false)
+        }
+    }
 
-        let callback = (reportProgress && Progress.current() != nil) ? Progress(totalUnitCount: Int64(buffer.count)) : nil
+    writeExtendedAttributes(fd: fd, attributes: attributes)
 
-        do {
-            try write(buffer: buffer, toFileDescriptor: fd, path: inPath, parentProgress: callback)
-        } catch {
-            try auxPath.withNTPathRepresentation { pwszAuxPath in
-                _ = DeleteFileW(pwszAuxPath)
-            }
+    _close(fd)
+    fd = -1
 
-            if callback?.isCancelled ?? false {
-                throw CocoaError(.userCancelled)
-            } else {
-                throw CocoaError.errorWithFilePath(inPath, errno: errno, reading: false)
-            }
+    try auxPath.withNTPathRepresentation { pwszAuxiliaryPath in
+        var hFile = CreateFileW(pwszAuxiliaryPath, DELETE,
+                                FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                                nil, OPEN_EXISTING,
+                                FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OPEN_REPARSE_POINT,
+                                nil)
+        if hFile == INVALID_HANDLE_VALUE {
+            let dwError = GetLastError()
+            _ = DeleteFileW(pwszAuxiliaryPath)
+            throw CocoaError.errorWithFilePath(inPath, win32: dwError, reading: false)
+        }
+        defer {
+            if hFile == INVALID_HANDLE_VALUE { return }
+            _ = CloseHandle(hFile)
         }
 
-        writeExtendedAttributes(fd: fd, attributes: attributes)
+        try inPath.path.withNTPathRepresentation { pwszPath in
+            let cbSize = wcslen(pwszPath) * MemoryLayout<WCHAR>.size
+            let dwSize = DWORD(MemoryLayout<FILE_RENAME_INFO>.size + cbSize)
```

Comment:
Here the allocation size for the name (`cbSize`) doesn't contain space for the terminating null byte, but below when we call `wcscpy_s` into `pInfo?.pointee.FileName` we provide the length `wcslen(pwszPath) + 1` to include the null terminating byte. Should `cbSize` have a `+ 1` to include the null byte in the allocation size? My rough read is that this over-writes the buffer by one scalar for the null byte.

Reply:
Ah, right, the count is off by 1, thanks!